packages/nimble-components/src/wafer-map/modules/rendering.ts

```diff
@@ -49,61 +64,76 @@ export class RenderingModule {
         return '';
     }
 
-    private sortDies(): void {
-        this.dies = this.wafermap.dataManager.diesRenderInfo.sort((a, b) => {
-            if (a.fillStyle > b.fillStyle) {
-                return 1;
-            }
-            if (b.fillStyle > a.fillStyle) {
-                return -1;
-            }
-
-            return 0;
-        });
-    }
-
     private renderDies(): void {
-        let prev: DieRenderInfo | undefined;
+        let fillStyle = '';
+        const context = this.wafermap.canvasContext;
+        const dieWidth = this.wafermap.dataManager.dieDimensions.width;
+        const dieHeight = this.wafermap.dataManager.dieDimensions.height;
+        const transformedCanvasMinPoint = this.wafermap.transform.invert([
+            0, 0
+        ]);
+        const transformedCanvasMaxPoint = this.wafermap.transform.invert([
+            this.wafermap.canvas.width,
+            this.wafermap.canvas.height
+        ]);
+        transformedCanvasMinPoint[0] -= dieWidth;
+        transformedCanvasMinPoint[1] -= dieHeight;
 
         for (const die of this.dies) {
-            if (!prev) {
-                this.wafermap.canvasContext.fillStyle = die.fillStyle;
-            }
-            if (prev && die.fillStyle !== prev.fillStyle && die.fillStyle) {
-                this.wafermap.canvasContext.fillStyle = die.fillStyle;
+            if (
+                this.isDieVisible(
+                    die,
+                    transformedCanvasMinPoint,
+                    transformedCanvasMaxPoint
+                )
+            ) {
+                if (fillStyle !== die.fillStyle) {

```

Comment:
Why bother with the `if` block here? The only possible difference the `if` block provides is if they were the same, but `context.fillStyle` _wasn't_ set to `die.fillStyle`, and you wanted to maintain that misalignment, which I'm guessing isn't the case.

Reply:
Setting the canvas context fillStyle is an intensive operation so we want to limit it as much as we can.
The dies are presorted based on their fill style because of that, so we want to change it only when a new batch of dies needs to be drawn

Reply:
Think it's worth adding a comment around any performance cliffs / surprising API perf behavior that you find, just a quick `Updating the canvas  context fillStyle may be slow so avoid if unnecessary`. A link to issues or discussion on StackOverflow, etc. with context could be nice if you have it.

Helpful for refactors etc to know what to double check for. Also those kind of perf cliffs may differ across browsers or platforms or go away over time so the context is useful.

Reply:
I rechecked the assumptions we had with an even larger 5M dataset and the repeated assignment of the canvas fillStyle only increased the runtime with 20 ms out of a 600 total, which is less than what we were fearing. I think the filtering reduced the impact it had in the past on performance.

Given that we plan to get rid of the merge sorting and replace it with a bucket sorting when creating the render info, i removed the if block altogether